### PR TITLE
CBG-584 Add overrides for ConnectBucket and Connect to remap hostnames based on AlternateAddress

### PR DESCRIPTION
**Pull request by bbrks**
Opening as draft to walk through the approach before tidying.

- [x] Requires http://review.couchbase.org/c/117745/

### CODE REVIEW

File: base/dcp_sharded.go [hunk 1]
```diff
@@ -118,12 +124,101 @@ func createCBGTIndex(manager *cbgt.Manager, dbName string, bucket Bucket, spec B
...
-	if spec.IsTLS() {
-		cbgt.RegisterBucketDataSourceOptionsCallback(indexName, manager.UUID(), func(options *cbdatasource.BucketDataSourceOptions) *cbdatasource.BucketDataSourceOptions {
-			options.TLSConfig = spec.TLSConfig
-			return options
-		})
-	}
+	cbgt.RegisterBucketDataSourceOptionsCallback(indexName, manager.UUID(), func(options *cbdatasource.BucketDataSourceOptions) *cbdatasource.BucketDataSourceOptions {
+		// A lookup of host dest to external alternate address hostnames
+		externalAlternateAddresses := make(map[string]string)
+		externalAlternateAddressesLock := sync.RWMutex{}
```
Comment (adamcfraser) [inline base/dcp_sharded.go]: Does the scoping prevent reusing the code for the non-sharded case here?  Perhaps a helper function that returns ConnectFunc and ConnectBucketFunc?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
The general approach seems viable, given that we're working with the restriction that we need to look up the alternate address given only the destination string passed to the custom Connect() implementation.

This use of the alternate host in Connect seems to follow the same pattern used by gocbcore in routeconfig.go. (I don't think gocb retrieves the alternate addresses via the pools endpoint - I think it uses cccp, but using pools makes sense given how go-couchbase manages a connection).

I noted a few questions on the specifics of the implementation - have a look and let me know what you think.

File: base/dcp_receiver.go [hunk 2]
```diff
@@ -205,6 +210,26 @@ func (nph NoPasswordAuthHandler) GetCredentials() (username string, password str
 	return "", "", bucketname
 }
 
+type bucketWrapper struct {
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: Opened Gerrit CR to add GetUUID() to couchbase.Bucket: http://review.couchbase.org/c/117813/

File: base/dcp_receiver.go [hunk 3]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
+
+					nodeHostname, _, err := SplitHostPort(node.Hostname)
+					if err != nil {
+						return nil, err
+					}
+
+					fmt.Printf("storing %s => %s\n", nodeHostname, external.Hostname+port)
+					externalAlternateAddressesLock.Lock()
+					externalAlternateAddresses[nodeHostname] = external.Hostname + port
+					externalAlternateAddressesLock.Unlock()
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: Nilling out the map for each call seems like the best approach here. Thanks!

File: base/dcp_sharded.go (hunk 1 above)
Comment (bbrks) [reply base/dcp_sharded.go]: There shouldn't be anything stopping me from sharing this code. I was just being lazy for the POC.

File: base/dcp_receiver.go [hunk 4]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
+			fmt.Printf("found alternate address for %s => %s\n", dest, extHostname)
+
+			_, port, _ := SplitHostPort(extHostname)
+			if port == "" {
+				port = destPort
+			}
+
+			dest = extHostname + ":" + port
+			fmt.Printf("using dest: %s\n", dest)
+		}
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: We'll continue to connect using the original hostname if no alternative is present. If that's not a routable hostname, we'll fail on `Connect`, and re-invoke `ConnectBucket` to repopulate the alt address map.

File: base/dcp_receiver.go [hunk 5]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
 		auth = NoPasswordAuthHandler{handler: spec.Auth}
 	}
 
-	// If using TLS, pass a custom connect method to support using TLS for cbdatasource's memcached connections
-	if spec.IsTLS() {
-		dataSourceOptions.Connect = spec.TLSConnect
+	// A lookup of host dest to external alternate address hostnames
+	externalAlternateAddresses := make(map[string]string)
+	externalAlternateAddressesLock := sync.RWMutex{}
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: I'm just unclear on the ConnectBucket/Connect timing.

It's not racy without the lock, but I've yet to try rebalancing a cluster. I'll test it without the locking to see if we can get away without it.

File: base/dcp_receiver.go [hunk 6]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
+		if spec.IsTLS() {
+			// If using TLS, pass a custom connect method to support using TLS for cbdatasource's memcached connections
+			return spec.TLSConnect(protocol, dest)
+		}
+
+		return memcached.Connect(protocol, dest)
+	}
+
+	// COPY OF cbdatasource.ConnectBucket
+	dataSourceOptions.ConnectBucket = func(serverURL, poolName, bucketName string, auth couchbase.AuthHandler) (cbdatasource.Bucket, error) {
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: Yeah. Upon rebalance, `ConnectBucket` is reinvoked.

File: base/dcp_receiver.go [hunk 7]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
+			pool, err := client.GetPool(poolName)
+			if err != nil {
+				return nil, err
+			}
+
+			// Build map of external alternate addresses if available
+			for _, node := range pool.Nodes {
+				if external, ok := node.AlternateNames["external"]; ok && external.Hostname != "" {
+					var port string
+					if extPort, ok := external.Ports["direct"]; ok {
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: This was wrong after digging through some obscure docs. Values should be "kv" and "kvSSL".

File: base/dcp_receiver.go [hunk 8]
```diff
@@ -284,9 +309,96 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
+			if port == "" {
+				port = destPort
+			}
+
+			dest = extHostname + ":" + port
+			fmt.Printf("using dest: %s\n", dest)
+		}
+		externalAlternateAddressesLock.RUnlock()
+
+		if spec.IsTLS() {
```
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: I've switched to using the TLSConfig option in this PR - see what you think.

File: base/dcp_receiver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbrks) [reply base/dcp_receiver.go]: I ran it under `-race` and at no point did anything get flagged. So I presume it's safe enough to remove the lock.

File: base/dcp_sharded.go [hunk 9]
```diff
@@ -3,10 +3,12 @@ package base
...
 	"strings"
 
-	"github.com/couchbase/cbgt"
-	"github.com/couchbase/go-couchbase/cbdatasource"
 	"github.com/pkg/errors"
 	pkgerrors "github.com/pkg/errors"
+
+	"github.com/couchbase/cbgt"
+
+	"github.com/couchbase/go-couchbase/cbdatasource"
```
Comment (adamcfraser) [inline base/dcp_sharded.go:9]: Import formatting cleanup here.

File: base/dcp_common.go [hunk 10]
```diff
@@ -536,3 +540,129 @@ func GenerateDcpStreamName(feedID string) (string, error) {
...
+			return nil, err
+		}
+
+		pool, err := client.GetPool(poolName)
+		if err != nil {
+			return nil, err
+		}
+
+		// Build map of external alternate addresses if available
+		externalAlternateAddresses = make(map[string]string, len(pool.Nodes))
```
Comment (adamcfraser) [inline base/dcp_common.go:603]: Isn't this map already initialized/allocated?  Do we need to do another make here? (alternately, do we need to be doing the make we're doing prior to calling alternateAddressShims)?

File: base/dcp_receiver.go [hunk 11]
```diff
@@ -284,11 +286,14 @@ func StartDCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, c
...
 
-	// If using TLS, pass a custom connect method to support using TLS for cbdatasource's memcached connections
-	if spec.IsTLS() {
-		dataSourceOptions.Connect = spec.TLSConnect
+	dataSourceOptions.TLSConfig = func() *tls.Config {
+		return spec.TLSConfig()
 	}
 
+	// A lookup of host dest to external alternate address hostnames
+	externalAlternateAddresses := make(map[string]string)
```
Comment (adamcfraser) [inline base/dcp_receiver.go]: I find the scoping for externalAlternateAddresses a bit difficult to reason about, but I'm assuming the reason your instantiating here (as opposed to in alternateAddressShims) is required for scoping?  If so, maybe stick a comment here with some details?

File: base/dcp_common.go [hunk 12]
```diff
@@ -536,3 +540,129 @@ func GenerateDcpStreamName(feedID string) (string, error) {
...
+		for _, node := range pool.Nodes {
+			if external, ok := node.AlternateNames["external"]; ok && external.Hostname != "" {
+				nodeHostname, _, err := SplitHostPort(node.Hostname)
+				if err != nil {
+					return nil, err
+				}
+
+				// TODO: check if host is tls
```
Comment (adamcfraser) [inline base/dcp_common.go]: This looks like it's still pending?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Generally looks good, and I like the shims approach.  A few comments for clarity, and what looks like a required TODO for kv vs kvssl.

File: base/dcp_receiver.go (hunk 11 above)
Comment (bbrks) [reply base/dcp_receiver.go]: I think this is an oversight from when the code used to live here actually. I'll retain the existing make inside the shim helper func.

File: base/dcp_common.go (hunk 12 above)
Comment (bbrks) [reply base/dcp_common.go]: oops, yeah - this TODO is associated with the `if true {` below - was intending to come back to this. I'll update tomorrow if I get chance.

File: base/dcp_common.go (hunk 10 above)
Comment (bbrks) [reply base/dcp_common.go:603]: The intention here was to remove any existing entries by recreating the map. I can add a comment to clarify.

**Comment by bbrks** [comment]
Addressed other comments but haven't got time right now to do the TLS check. I'll come back to it tomorrow.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**